test/Npgsql.Tests/ConnectionTests.cs

```diff
@@ -1559,6 +1559,102 @@ public async Task TimeoutDuringAuthentication()
         }
 
+        [Test]
+        public async Task Physical_open_callback_sync()
+        {
+            await using var defaultConn = await OpenConnectionAsync();
+            await using var _ = await CreateTempTable(defaultConn, "ID INTEGER", out var table);
+
+            using var __ = CreateTempPool(ConnectionString, out var connectionString);
+            using var conn = new NpgsqlConnection(connectionString);
+            conn.PhysicalOpenCallback = connector =>
+            {
+                using var cmd = connector.CreateCommand($"INSERT INTO \"{table}\" VALUES(1)");
+                cmd.ExecuteNonQuery();
+                return new ValueTask();
+            };
+
+            conn.Open();
+
+            var rowsCount = (long)(await conn.ExecuteScalarAsync($"SELECT COUNT(*) FROM \"{table}\""))!;
+            Assert.AreEqual(1, rowsCount);
+        }
+
+        [Test]
+        public async Task Physical_open_callback_async()
+        {
+            await using var defaultConn = await OpenConnectionAsync();
+            await using var _ = await CreateTempTable(defaultConn, "ID INTEGER", out var table);
+
+            using var __ = CreateTempPool(ConnectionString, out var connectionString);
+            await using var conn = new NpgsqlConnection(connectionString);
+            conn.PhysicalOpenCallback = async connector =>
+            {
+                using var cmd = connector.CreateCommand($"INSERT INTO \"{table}\" VALUES(1)");
+                await cmd.ExecuteNonQueryAsync();
+            };
+
+            await conn.OpenAsync();
+
+            var rowsCount = (long)(await conn.ExecuteScalarAsync($"SELECT COUNT(*) FROM \"{table}\""))!;
+            Assert.AreEqual(1, rowsCount);
+        }
+
+        [Test]
+        public async Task Physical_open_callback_async_over_sync()
+        {
+            using var _ = CreateTempPool(ConnectionString, out var connectionString);
+            await using var conn = new NpgsqlConnection(connectionString);
+            conn.PhysicalOpenCallback = async connector =>
+            {
+                await Task.Delay(50);
+            };
+
+            Assert.Throws<InvalidOperationException>(conn.Open);
+        }
+
+        [Test]
+        public async Task Physical_open_callback_throws()
+        {
+            PhysicalOpenCallback callback = _ => throw new NotImplementedException();
+
+            using var _ = CreateTempPool(ConnectionString, out var connectionString);
+            await using var conn = new NpgsqlConnection(connectionString);
+            conn.PhysicalOpenCallback = callback;
+
+            Assert.ThrowsAsync<NotImplementedException>(conn.OpenAsync);
+
+            if (IsMultiplexing)
+            {
+                await conn.CloseAsync();
+                conn.PhysicalOpenCallback = null;
+                // Allow the bootstrap to complete
+                Assert.DoesNotThrowAsync(conn.OpenAsync);
+
+                NpgsqlConnection.ClearPool(conn);
```

Comment:
What's the purpose of this extra test code for multiplexing? It seems to clear the pool and repeat the test just without the bootstrap, which doesn't seem related to this callback... Is there a particular thing you're trying to check here?

Reply:
> Is there a particular thing you're trying to check here?

Yep - with multiplexing, the physical connection might open in 2 places: `NpgsqlConnection.Open` (if there was no bootstrap beforehand), and `NpgsqlCommand.Execute`. This is testing the second case.